File: key-value/python/hashmap.py


import time
import random
import math
from typing import List

class Slot:
	def __init__(self, key: str, value: str, next_: "Slot") -> None:
		self.key: str = key
		self.value: str = value
		self.timestamp: float = time.time()
		self.next_: "Slot" = next_

	def set(self, value: str) -> None:
		self.value = value
		self.timestamp = time.time()

	def get(self) -> str:
		self.timestamp = time.time()
		return self.value

class PairSlotIndex:
	def __init__(self, s: Slot, i: int) -> None:
		self.slot: Slot = s
		self.index: int = i

class EvictionPoolEntry:
	def __init__(self, s: Slot) -> None:
		self.key: str = s.key
		self.timestamp: float = s.timestamp

class EvictionPool:
	MAX_LENGTH: int = 16
	CANDIDATE_COUNT: int = 5

	def __init__(self, hashmap: "HashMap") -> None:
		self.hashmap: "HashMap" = hashmap
		self.pool: List[EvictionPoolEntry] = [None] * EvictionPool.MAX_LENGTH
		random.seed(322)

	def populate(self) -> None:
		candidates: List[Slot] = [None] * EvictionPool.CANDIDATE_COUNT
		i: int = 0
		while i < len(candidates):
			row: int = math.floor(random.random() * len(self.hashmap.rows))
			s: Slot = self.hashmap.rows[row]
			while s != None and i < len(candidates):
				candidates[i] = s
				i += 1
				s = s.next_

		for candidate in candidates:
			for j in range(len(self.pool)):
				if self.pool[j] == None:
					self.pool[j] = EvictionPoolEntry(candidate)
					break
				elif self.pool[j].timestamp < candidate.timestamp:
					self.pool[j + 1:] = self.pool[j:-1]
					self.pool[j] = EvictionPoolEntry(candidate)
					break

	def pop(self) -> str:
		k: str = None
		if self.pool[0] != None:
			k = self.pool[0].key
			self.pool[:] = self.pool[1:] + [None]
		return k


class HashMap:
	CANDIDATE_COUNT: int = 5

	def __init__(self, max_size: int, num_rows: int) -> None:
		self.rows: List[Slot] = [None] * num_rows
		self.size: int = 0
		self.eviction_pool: EvictionPool = EvictionPool(self)
		self.max_size: int = max_size

	@staticmethod
	def hash(s: str) -> int:
		h: int = 0
		for i in range(len(s)):
			h = (h * 31) + ord(s[i])
		return h

	def find(self, key: str) -> PairSlotIndex:
		i: int = HashMap.hash(key) % len(self.rows)
		s: Slot = self.rows[i]
		while s != None:
			if s.key == key:
				return PairSlotIndex(s, i)
			s = s.next_
		return PairSlotIndex(None, i)

	def exists(self, key: str) -> bool:
		return True if self.find(key).slot != None else False

	def get(self, key: str) -> str:
		p: PairSlotIndex = self.find(key)
		if p.slot == None:
			return None
		return p.slot.get()

	def set(self, key: str, value: str) -> int:
		if self.size > self.max_size:
			self.evict()

		p: PairSlotIndex = self.find(key)
		if p.slot != None:
			self.size += len(value) - len(p.slot.value)
			p.slot.set(value)
			return 1

		self.size += len(value)
		s: Slot = Slot(key, value, self.rows[p.index])
		self.rows[p.index] = s
		return 0

	class DelPair:
		def __init__(self) -> None:
			self.keys: int = 0
			self.size: int = 0

	def delete(self, key: str) -> "HashMap.DelPair":
		ret: "HashMap.DelPair" = HashMap.DelPair()
		i: int = HashMap.hash(key) % len(self.rows)
		prev: Slot = None
		s: Slot = self.rows[i]
		while s != None:
			if s.key == key:
				if prev == None:
					self.rows[i] = s.next_
				else:
					prev.next_ = s.next_
				ret.keys += 1
				ret.size = len(s.value)
				self.size -= ret.size
				break
			prev = s
			s = s.next_
		return ret

	def sample(self) -> str:
		key: str = None
		oldest: int = 0
		i: int = 0
		while i < HashMap.CANDIDATE_COUNT:
			row: int = math.floor(random.random() * len(self.rows))
			s: Slot = self.rows[row]
			while s != None and i < HashMap.CANDIDATE_COUNT:
				if oldest == 0 or oldest > s.timestamp:
					key = s.key
					oldest = s.timestamp
				i += 1
				s = s.next_
		return key

	def evict(self) -> int:
		size_freed: int = 0
		while self.size > self.max_size:
			key: str = self.sample()
			size_freed += self.delete(key).size
		return size_freed

File: key-value/go/hashmap.go
package main

import "sync"
import "sync/atomic"
import "math/rand"
import "time"

func assert(cond bool) {
	if !cond {
		panic("badness")
	}
}

type Row struct {
	head *Slot
	lock sync.RWMutex
}

type Slot struct {
	key string
	value string
	timestamp int64
	next *Slot
}

func newSlot(key string, value string, next *Slot) *Slot {
	var s *Slot = new(Slot)
	s.key = key
	s.value = value
	s.next = next
	s.timestamp = time.Now().UnixNano()
	return s
}

func (this *Slot) set(value string) {
	this.value = value
	this.timestamp = time.Now().UnixNano()
}

func (this *Slot) get() *string {
	this.timestamp = time.Now().UnixNano()
	return &this.value
}

type EvictionPoolEntry struct {
	key string
	timestamp int64
}

func newEvictionPoolEntry(s *Slot) *EvictionPoolEntry {
	var epe *EvictionPoolEntry = new(EvictionPoolEntry)
	epe.key = s.key
	epe.timestamp = s.timestamp
	return epe
}

const MAX_LENGTH int = 16
const CANDIDATE_COUNT int = 5
type EvictionPool struct {
	pool []*EvictionPoolEntry
	hashmap *HashMap
}

type HashMap struct {
	max_size int64
	rows []Row
	length int
	size int64 // atomic
}

func NewHashMap(max_size int64, num_rows int) *HashMap {
	var hm *HashMap = new(HashMap)
	hm.max_size = max_size
	hm.length = num_rows
	hm.rows = make([]Row, hm.length)
	return hm
}

func hash(key string) int64 {
	var h int64 = 0
	for i := 0; i < len(key); i++ {
		h = (h * 31) + int64(key[i])
	}
	return h
}

func (this *HashMap) get_row(key string) int {
	return (int(hash(key) % int64(this.length)) + this.length) % this.length
}

func (this *HashMap) find(key string, row int) *Slot {
	for s := this.rows[row].head; s != nil; s = s.next {
		if s.key == key {
			return s
		}
	}
	return nil
}

func (this *HashMap) exists(key string) bool {
	var i int = this.get_row(key)
	this.rows[i].lock.RLock()
	defer this.rows[i].lock.RUnlock()
	var s *Slot = this.find(key, i)
	return (s != nil)
}

type DelPair struct {
	count int
	bytes int
}

func (this *HashMap) del(key string) DelPair {
	var ret DelPair = DelPair { 0, 0 }
	var prev *Slot = nil
	var i int = this.get_row(key)
	this.rows[i].lock.Lock()
	defer this.rows[i].lock.Unlock()

	for s := this.rows[i].head; s != nil; s = s.next {
		if s.key == key {
			if prev == nil {
				this.rows[i].head = s.next
			} else {
				prev.next = s.next
			}
			ret.count++;
			ret.bytes = len(s.value)
			atomic.AddInt64(&this.size, int64(-ret.bytes))
			break
		}
		prev = s
	}
	return ret
}

func (this *EvictionPool) populate() {
	var candidates []*Slot = make([]*Slot, CANDIDATE_COUNT)
	for i := 0; i < CANDIDATE_COUNT; {
		var row int = ((rand.Int() % this.hashmap.length) + this.hashmap.length) % this.hashmap.length
		{
			this.hashmap.rows[row].lock.RLock()
			for s := this.hashmap.rows[row].head; s != nil && i < CANDIDATE_COUNT; s = s.next {
				candidates[i] = s
				i++
			}
			this.hashmap.rows[row].lock.RUnlock()
		}
	}

	// insert the candidates
	for i := 0; i < CANDIDATE_COUNT; i++ {
		var candidate *Slot = candidates[i]
		for j := 0; j < MAX_LENGTH; j++ {
			if this.pool[j] == nil {
				this.pool[j] = newEvictionPoolEntry(candidate)
				break
			} else if this.pool[j].timestamp < candidate.timestamp {
				for k := MAX_LENGTH - 1; k > j; k-- {
					this.pool[k] = this.pool[k - 1]
				}
				this.pool[j] = newEvictionPoolEntry(candidate)
				break
			}
		}
	}
}

func (this *EvictionPool) pop() *string {
	var ret *string = nil
	if this.pool[0] != nil {
		ret = &this.pool[0].key
		for i := 0; i < MAX_LENGTH - 1; i++ {
			this.pool[i] = this.pool[i + 1]
		}
		this.pool[MAX_LENGTH - 1] = nil
	}
	return ret
}

func (this *HashMap) sample() string {
	var key string
	var oldest int64 = 0
	for i := 0; i < CANDIDATE_COUNT; {
		var row int = ((rand.Int() % this.length) + this.length) % this.length
		{
			this.rows[row].lock.RLock()
			for s := this.rows[row].head; s != nil && i < CANDIDATE_COUNT; s = s.next {
				if oldest == 0 || oldest > s.timestamp {
					key = s.key
					oldest = s.timestamp
				}
			}
			i++;
			this.rows[row].lock.RUnlock()
		}
	}
	return key
}

func (this *HashMap) evict() int {
	var size_freed int = 0
	for atomic.LoadInt64(&this.size) > this.max_size {
		var key string = this.sample()
		size_freed += this.del(key).bytes
	}
	return size_freed
}

func (this *HashMap) Get(key string) *string {
	var i int = this.get_row(key)
	this.rows[i].lock.RLock()
	defer this.rows[i].lock.RUnlock()
	var s *Slot = this.find(key, i)
	if s == nil {
		return nil
	}
	return s.get()
}

func (this *HashMap) Set(key string, value string) int {
	if atomic.LoadInt64(&this.size) > this.max_size {
		this.evict();
	}
	var i int = this.get_row(key)
	this.rows[i].lock.Lock()
	defer this.rows[i].lock.Unlock()
	var s *Slot = this.find(key, i)
	if s != nil {
		atomic.AddInt64(&this.size, int64(len(value) - len(s.value)))
		s.set(value)
		return 1
	}
	atomic.AddInt64(&this.size, int64(len(value)))
	s = newSlot(key, value, this.rows[i].head)
	this.rows[i].head = s
	return 0
}
